File: I1_MVPAndRetrofit/app/src/main/java/com/example/tsvetelinastoyanova/weatherreportrevisited/data/source/CitiesRepository.java
package com.example.tsvetelinastoyanova.weatherreportrevisited.data.source;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.example.tsvetelinastoyanova.weatherreportrevisited.City;
import com.example.tsvetelinastoyanova.weatherreportrevisited.data.CityEntity;
import com.example.tsvetelinastoyanova.weatherreportrevisited.data.source.local.LocalDataSource;
import com.example.tsvetelinastoyanova.weatherreportrevisited.data.source.remote.CitiesRemoteDataSource;
import com.example.tsvetelinastoyanova.weatherreportrevisited.model.WeatherObject;
import com.example.tsvetelinastoyanova.weatherreportrevisited.util.CityEntityAdapter;
import com.example.tsvetelinastoyanova.weatherreportrevisited.util.Utils;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

public class CitiesRepository implements CityDataSource {
    private static CitiesRepository INSTANCE = null;
    private final LocalDataSource localDataSource;
    private final CitiesRemoteDataSource remoteDataSource;

    List<WeatherObject> weatherObjectsCache = new ArrayList<>();

    private CitiesRepository(@NonNull LocalDataSource localDataSource,
                             @NonNull CitiesRemoteDataSource remoteDataSource) {
        this.localDataSource = checkNotNull(localDataSource);
        this.remoteDataSource = checkNotNull(remoteDataSource);
    }

    public static CitiesRepository getInstance(LocalDataSource localDataSource,
                                               CitiesRemoteDataSource remoteDataSource) {
        if (INSTANCE == null) {
            INSTANCE = new CitiesRepository(localDataSource, remoteDataSource);
        }
        return INSTANCE;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }

//    public List<WeatherObject> getWeatherObjectsCache() {
//        return weatherObjectsCache;
//    }

    public void getCities(@NonNull CityDataSource.GetCityCallback getCityCallback) {
        /*
        if dirty cache:
        1) getAll cities from DB
                1.1) success - forEach cityEntity request to the API
                    1.1.1) success - return list of cityEntities
                    1.1.2) fail - error with internet
                1.2) fail - error
         else:
         2) getAll cities from DB
                2.1) success - return cityEntities from DB one by one
                2.2) fail - error
         */
        Utils.checkNotNull(getCityCallback);
        if (!weatherObjectsCache.isEmpty()) {
            getCitiesFromCache(getCityCallback);
        } else {
            getCitiesFromDataSource(getCityCallback);
        }
    }

    private void getCitiesFromDataSource(@NonNull GetCityCallback getCityCallback) {
        localDataSource.getCities(new LocalDataSource.LoadCitiesCallback() {
            @Override
            public void onCitiesLoaded(List<CityEntity> cities) {
                if (!cities.isEmpty()) {
                    for (CityEntity cityEntity : cities) {
                        remoteDataSource.getWeatherObject(cityEntity.getName(), new GetWeatherObjectCallback() {
                            @Override
                            public void onWeatherObjectLoaded(WeatherObject weatherObject) {
                                refreshWeatherObjectInCache(weatherObject);
                                getCityCallback.onCityLoaded(CityEntityAdapter.convertWeatherObjectToCityEntity(weatherObject));
                            }

                            @Override
                            public void onFail() {
                                getCityCallback.onCityDoesNotExist();
                            }
                        });
                    }
                }
            }

            @Override
            public void onDataNotAvailable() {
                getCityCallback.onCityDoesNotExist();
            }
        });
    }

    private void getCitiesFromCache(@NonNull GetCityCallback getCityCallback) {
        synchronized (weatherObjectsCache) {
            for (WeatherObject weatherObject : weatherObjectsCache) {
                CityEntity cityEntity = CityEntityAdapter.convertWeatherObjectToCityEntity(weatherObject);
                getCityCallback.onCityLoaded(cityEntity);
            }
        }
    }

    @Override
    public void getCity(@NonNull String cityName, @NonNull GetCityCallback callback) {
        Utils.checkNotNull(cityName);
        Utils.checkNotNull(callback);

        localDataSource.getCity(cityName, new GetCityCallback() {
            @Override
            public void onCityLoaded(CityEntity city) {
                callback.onCityLoaded(city);
            }

            @Override
            public void onCityDoesNotExist() {
                callback.onCityDoesNotExist();
            }
        });
    }

    public void addCity(String cityName, @NonNull AddCityCallback callback) {
        /* 1) check if city exists in database
            1.1) exists - onCityAlreadyExists()
            1.2) does not exist - api request()
                1.2.1) success
                    1.2.1.1) write to DB
                        1.2.1.1.1) success - return message
                        1.2.1.1.2) fail - return message
                1.2.2) fail - return message */

        Utils.checkNotNull(cityName);
        Utils.checkNotNull(callback);

        localDataSource.getCity(cityName, new GetCityCallback() {
            @Override
            public void onCityDoesNotExist() {
                remoteDataSource.getWeatherObject(cityName, new GetWeatherObjectCallback() {
                    @Override
                    public void onWeatherObjectLoaded(WeatherObject weatherObject) {
                        localDataSource.addCity(CityEntityAdapter.convertWeatherObjectToCityEntity(weatherObject), new LocalDataSource.AddCityCallback() {
                            @Override
                            public void onCityAddedSuccessfully(CityEntity cityEntity) {
                                refreshWeatherObjectInCache(weatherObject);
                                callback.onCityAddedSuccessfully(cityEntity);
                            }

                            @Override
                            public void onFail() {
                                callback.onFail();
                            }
                        });
                    }

                    @Override
                    public void onFail() {
                        callback.onFail();
                    }
                });
            }

            @Override
            public void onCityLoaded(CityEntity city) {
                callback.onFail();
            }
        });
    }

    public void deleteCity(@NonNull String cityName, LocalDataSource.DeleteCityCallback deleteCityCallback) {
        Utils.checkNotNull(cityName);
        localDataSource.deleteCity(cityName, new LocalDataSource.DeleteCityCallback() {
            @Override
            public void onCityDeletedSuccessfully() {
                deleteWeatherObjectFromCache(cityName);
                deleteCityCallback.onCityDeletedSuccessfully();
            }

            @Override
            public void onFail() {
                deleteCityCallback.onFail();
            }
        });
    }

    public WeatherObject getWeatherObjectWithName(String cityName) {
        synchronized (weatherObjectsCache) {
            int saveIndex;
            for (WeatherObject weatherObject: weatherObjectsCache) {
                if (weatherObject.getName().equals(cityName)) {
                    return weatherObject;
                }
            }
            saveIndex = weatherObjectsCache.size();

            addWeatherObjectToCacheFromRemote(cityName);

            if (saveIndex > weatherObjectsCache.size()) {
                for (int i = saveIndex - 1; i < weatherObjectsCache.size(); i++) {
                    if (weatherObjectsCache.get(i).getName().equals(cityName)) {
                        return weatherObjectsCache.get(i);
                    }
                }
            }
        }
        return null;
    }

    public void refreshCities(List<City> cities, @NonNull CityDataSource.GetCityCallback getCityCallback) {
        /* foreach city
            load from DB
            get name and load from API
            refresh cache
            refresh DB
            refresh UI
         */
        for (City c : cities) {
            getCity(c.getName(), new GetCityCallback() {
                @Override
                public void onCityLoaded(CityEntity oldCity) {
                    remoteDataSource.getWeatherObject(oldCity.getName(), new GetWeatherObjectCallback() {
                        @Override
                        public void onWeatherObjectLoaded(WeatherObject weatherObject) {
                            refreshWeatherObjectInCache(weatherObject);
                            CityEntity newCity = CityEntityAdapter.convertWeatherObjectToCityEntity(weatherObject);
                            localDataSource.refreshCity(newCity, (() -> getCityCallback.onCityLoaded(newCity)));
                        }

                        @Override
                        public void onFail() {
                            getCityCallback.onCityDoesNotExist();
                        }
                    });
                }

                @Override
                public void onCityDoesNotExist() {
                    getCityCallback.onCityDoesNotExist();
                }
            });
        }
    }

    public void refreshWeatherObjectInCache(WeatherObject newObject) {
        synchronized (this) {
            for (int i = 0; i < weatherObjectsCache.size(); i++) {
                if (weatherObjectsCache.get(i).getName().equals(newObject.getName())) {
                    weatherObjectsCache.set(i, newObject);
                    return;
                }
            }

            weatherObjectsCache.add(newObject);
        }
    }

    private void deleteWeatherObjectFromCache(String cityName) {
        synchronized (weatherObjectsCache) {
            for (WeatherObject weatherObject : weatherObjectsCache) {
                if (weatherObject.getName().equals(cityName)) {
                    weatherObjectsCache.remove(weatherObject);
                    return;
                }
            }
        }
    }

    private void addWeatherObjectToCacheFromRemote(String cityName) {
        synchronized (weatherObjectsCache) {
            remoteDataSource.getWeatherObject(cityName, new GetWeatherObjectCallback() {
                @Override
                public void onWeatherObjectLoaded(WeatherObject weatherObject) {
                    weatherObjectsCache.add(weatherObject);
                }

                @Override
                public void onFail() {

                }
            });
        }
    }
}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/data/source/CitiesRepository.kt
package com.example.tsvetelinastoyanova.weatherapp.data.source

import android.support.annotation.VisibleForTesting
import com.example.tsvetelinastoyanova.weatherapp.City
import com.example.tsvetelinastoyanova.weatherapp.data.CityEntity
import com.example.tsvetelinastoyanova.weatherapp.data.source.local.LocalDataSource
import com.example.tsvetelinastoyanova.weatherapp.data.source.remote.CitiesRemoteDataSource
import com.example.tsvetelinastoyanova.weatherapp.model.currentweather.CurrentWeatherObject
import com.example.tsvetelinastoyanova.weatherapp.model.forecast.ForecastObject
import com.example.tsvetelinastoyanova.weatherapp.util.convertToWeatherObjectWithCelsiusTemperature
import com.example.tsvetelinastoyanova.weatherapp.util.convertWeatherObjectToCityEntity
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.*

sealed class CityState
object ExistsInDatabase : CityState()
object NotValid : CityState()
data class AddedSuccessfully(val currentWeatherObject: CurrentWeatherObject, val cityEntity: CityEntity) : CityState()


open class CitiesRepository(private val localDataSource: LocalDataSource,
                            private val remoteDataSource: CitiesRemoteDataSource) : CityDataSource {

    private var weatherObjectsCache: MutableList<CurrentWeatherObject> = ArrayList()
    private var forecastObjectsCache: MutableList<ForecastObject> = ArrayList()

    companion object {
        @Volatile
        private var INSTANCE: CitiesRepository? = null

        fun getInstance(localDataSource: LocalDataSource,
                        remoteDataSource: CitiesRemoteDataSource): CitiesRepository {
            return INSTANCE ?: synchronized(this) {
                INSTANCE
                    ?: CitiesRepository(localDataSource, remoteDataSource).also { INSTANCE = it }
            }
        }
    }

    @VisibleForTesting
    open fun getSizeOfCache(): Int {
        return weatherObjectsCache.size
    }

    open fun getCities(getCityCallback: CityDataSource.GetCityCallback) {
        if (getSizeOfCache() > 0) {
            getCitiesFromCache(getCityCallback)
        } else {
            getCitiesFromDataSource(getCityCallback)
        }
    }

    open fun getCitiesFromDataSource(getCityCallback: CityDataSource.GetCityCallback) {
        localDataSource.getCities()
            .filter { list -> list.isNotEmpty() }
            .flatMapPublisher { listOfCityEntities ->
                Flowable.fromIterable(listOfCityEntities)
            }
            .flatMapSingle { cityEntity: CityEntity ->
                remoteDataSource.getWeatherObject(cityEntity.name)
            }
            .map { weatherObject ->
                weatherObject.convertToWeatherObjectWithCelsiusTemperature()
            }
            .subscribe(
                { currentWeatherObject ->
                    refreshWeatherObjectInCache(currentWeatherObject)
                    getCityCallback.onCityLoaded(convertWeatherObjectToCityEntity(currentWeatherObject))
                },
                { _ -> getCityCallback.onCityDoesNotExist() }
            )
    }

    private fun getCitiesFromCache(getCityCallback: CityDataSource.GetCityCallback) {
        synchronized(weatherObjectsCache) {
            for (weatherObject in weatherObjectsCache) {
                val cityEntity = convertWeatherObjectToCityEntity(weatherObject)
                getCityCallback.onCityLoaded(cityEntity)
            }
        }
    }

    override fun getCity(cityName: String): Single<CityEntity> {
        return localDataSource.getCity(cityName)
    }

    open fun addNotAddedCity(cityName: String, callback: LocalDataSource.AddCityCallback) {
        /*** 1) check if city exists in database
        1.1) exists - onCityAlreadyExists()
        1.2) does not exist - api request()
        1.2.1) success
        1.2.1.1) write to DB
        1.2.1.1.1) success - return message
        1.2.1.1.2) fail - return message
        1.2.2) fail - return message */

        val onErrorRecoverFlowable: Single<CityState> = remoteDataSource.getWeatherObject(cityName)
            .map { weatherObject ->
                weatherObject.convertToWeatherObjectWithCelsiusTemperature()
            }
            .flatMap { weatherObject ->
                addCityInLocalDataSource(weatherObject)
            }
            .map { (cityEntity, currentWeatherObject) ->
                AddedSuccessfully(currentWeatherObject, cityEntity)
            }

        localDataSource.getCity(cityName)
            .map<CityState> { ExistsInDatabase }
            //.doOnSuccess { _ -> callback.onCityExistsInDatabase() }
            .onErrorResumeNext(onErrorRecoverFlowable)
            .subscribe(
                { result ->
                    when (result) {
                        is AddedSuccessfully -> {
                            refreshWeatherObjectInCache(result.currentWeatherObject)
                            callback.onCityAddedSuccessfully(result.cityEntity)
                        }
                        is ExistsInDatabase -> callback.onCityExistsInDatabase()
                    }
                },
                { _ -> callback.onNotValidCity() }
            )
    }

    private fun addCityInLocalDataSource(weatherObject: CurrentWeatherObject) =
        localDataSource.addCity(convertWeatherObjectToCityEntity(weatherObject))
            .map { cityEntity: CityEntity -> cityEntity to weatherObject }


    open fun deleteCity(cityName: String, deleteCityCallback: LocalDataSource.DeleteCityCallback) {
        localDataSource.deleteCity(cityName)
            .doOnError { _ -> deleteCityCallback.onFail() }
            .map { delete -> deleteWeatherObjectFromCache(delete) }
            .subscribe(
                { _ -> deleteCityCallback.onCityDeletedSuccessfully() },
                { _ -> deleteCityCallback.onFail() }
            )
    }

    open fun getWeatherObjectWithName(cityName: String): CurrentWeatherObject? {
        synchronized(weatherObjectsCache) {
            for (weatherObject in weatherObjectsCache) {
                weatherObject.let {
                    if (it.name == cityName) {
                        return weatherObject
                    }
                }
            }
            val saveIndex = weatherObjectsCache.size
            addWeatherObjectToCacheAndDatabaseFromRemote(cityName)

            if (saveIndex > weatherObjectsCache.size) {
                for (i in saveIndex - 1 until weatherObjectsCache.size) {
                    if (weatherObjectsCache[i].name == cityName) {
                        return weatherObjectsCache[i]
                    }
                }
            }
        }
        return null
    }

    open fun refreshCities(cities: List<City>, getCityCallback: CityDataSource.GetCityCallback) {
        /*** foreach city
         * load from DB
         * get name and load from API
         * refresh cache
         * refresh DB
         * refresh UI
         ***/

        Flowable.fromIterable(cities)
            .flatMapSingle { city -> getCity(city.name) }
            .flatMapSingle { city -> remoteDataSource.getWeatherObject(city.name) }
            .map { weatherObject -> weatherObject.convertToWeatherObjectWithCelsiusTemperature() }
            .map { weatherObject ->
                refreshWeatherObjectInCache(weatherObject)
                weatherObject
            }
            .map { newCity -> convertWeatherObjectToCityEntity(newCity) }
            .flatMapSingle { newCity -> localDataSource.refreshCity(newCity) }
            .subscribe(
                { success -> getCityCallback.onCityLoaded(success) },
                { _ -> getCityCallback.onCityDoesNotExist() }
            )
    }

    fun getForecastForCity(city: String): Single<ForecastObject> {
        return if (!forecastObjectsCache.isEmpty()) {
            getForecastFromCache(city)
        } else {
            getForecastFromDataSource(city)
        }
    }

    fun addWeatherObjectToCacheAndDatabaseFromRemote(cityName: String) {
        synchronized(weatherObjectsCache) {
            remoteDataSource.getWeatherObject(cityName)
                .doOnSuccess { weatherObject -> localDataSource.addCity(convertWeatherObjectToCityEntity(weatherObject)) }
                .subscribe { weatherObject -> weatherObjectsCache.add(weatherObject) }
        }
    }

    private fun getForecastFromCache(city: String): Single<ForecastObject> {
        val forecastObject = forecastObjectsCache.find { s -> s.city.name == city }
        forecastObject ?: return getForecastFromDataSource(city)
        return Single.just(forecastObject)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
    }

    private fun getForecastFromDataSource(city: String): Single<ForecastObject> {
        return remoteDataSource.getForecastForCity(city)
            .map { newObject ->
                addForecastObjectInCache(newObject)
                newObject
            }
    }

    private fun addForecastObjectInCache(newObject: ForecastObject) {
        synchronized(this) {
            forecastObjectsCache.add(newObject)
        }
    }

    private fun refreshWeatherObjectInCache(newObject: CurrentWeatherObject) {
        synchronized(this) {
            var exists = false
            loop@ for (i in weatherObjectsCache.indices) {
                if (weatherObjectsCache[i].name == newObject.name) {
                    weatherObjectsCache[i] = newObject
                    exists = true
                    break@loop
                }
            }
            if (!exists) {
                weatherObjectsCache.add(newObject)
            }
        }
    }

    private fun deleteWeatherObjectFromCache(cityName: String) {
        synchronized(weatherObjectsCache) {
            loop@ for (weatherObject in weatherObjectsCache) {
                if (weatherObject.name == cityName) {
                    weatherObjectsCache.remove(weatherObject)
                    break@loop
                }
            }
        }
    }
}